OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -295,8 +315,14 @@ public virtual IEnumerable<ActorInfo> BuildableItems()
 				return Enumerable.Empty<ActorInfo>();
 			if (developerMode.AllTech)
 				return Producible.Keys;
-
-			return buildableProducibles;
+			if (!Info.PayUpFront || !Info.BlockedIfInsolvent)
+			{
+				return buildableProducibles;
+			}
+			else
+			{
+				return buildableProducibles.Where(a => GetProductionCost(a) < playerResources.GetCashAndResources() || IsInQueue(a));
+			}
```

Comment:
This also is a bit incorrect as the above developerMode.AllTech ignores the filtering. Please also filter `Producible.Keys` when `developerMode.AllTech` is true

Reply:
> Also why do we need to have separate `Info.BlockedIfInsolvent` here? Shouldn't this behaviour be the default for `PauUpFront`?

It was. But @MustaphaTR wanna it this way.

Reply:
> Also why do we need to have separate `Info.BlockedIfInsolvent` here? Shouldn't this behaviour be the default for `PauUpFront`?



> Also why do we need to have separate `Info.BlockedIfInsolvent` here? Shouldn't this behaviour be the default for `PauUpFront`?
> 
> If so, you can remove the change to ProductionPaletteWidget as we would already filter buildable items here.



> Also why do we need to have separate `Info.BlockedIfInsolvent` here? Shouldn't this behaviour be the default for `PauUpFront`?
> 
yes, only reason BlockedIfInsolvent was added, because that's how its behave in OG starport. But I also think item should be blocked if cannot be purchase So if there is consensus on this will gladly remove BlockedIfInsolvent .

Reply:
> If so, you can remove the change to ProductionPaletteWidget as we would already filter buildable items here.

Change in ProductionPaletteWidget  are to block click sound if item is building and you cannot queue more. It has nothing to do with `PauUpFront`

Reply:
> Change in ProductionPaletteWidget are to block click sound if item is building and you cannot queue more. It has nothing to do with `PayUpFront`

it does because the widget gets the building from `BuildableItems()`. A lot of things get it from there hence why I'm fairly afraid of regressions here. The name `BlockedIfInsolvent` tbf also doesn't make sense as the production is blocked regardless of what it is set to.

Reply:
> . The name `BlockedIfInsolvent` tbf also doesn't make sense as the production is blocked regardless of what it is set to.

What about `BlockIconIfInsolvent`?

Reply:
> A lot of things get it from there hence why I'm fairly afraid of regressions here

I test this feature on all ProductionQueue traits, with bots. Didnt found any regression. Only problem I found is that `ProductionBar` use Item cost instead of Ticks for bar calculation. I will address that in separate PR.